client/src/main/java/io/pravega/client/security/auth/JwtTokenProviderImpl.java

```diff
@@ -0,0 +1,234 @@
+               if (expiryTimeFieldParts != null && expiryTimeFieldParts.length == 2) {
+                   result = Long.parseLong(expiryTimeFieldParts[1].trim());
+               }
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Returns the delegation token. It returns existing delegation token if it is not close to expiry. If the token
+     * is close to expiry, it obtains a new delegation token and returns that one instead.
+     *
+     * @return String the delegation token JWT compact value
+     */
+    @Override
+    public String retrieveToken() {
+        if (this.delegationToken.get() == null) {
+            return this.refreshToken();
+        } else if (isTokenNearingExpiry()) {
+            log.info("Token is nearing expiry");
+            return refreshToken();
+        } else {
+            return delegationToken.get().getValue();
+        }
+    }
+
+    @Override
+    public boolean populateToken(@NonNull String token) {
+        if (token.trim().equals("")) {
+            return false;
+        } else {
+            this.delegationToken.set(new DelegationToken(token, extractExpirationTime(token)));
+            return true;
+        }
+    }
+
+    private boolean isTokenNearingExpiry() {
+        Long currentTokenExpirationTime = this.delegationToken.get().getExpiryTime();
+
+        // currentTokenExpirationTime can be null if the server returns a null delegation token
+        return currentTokenExpirationTime != null && isWithinRefreshThreshold(currentTokenExpirationTime);
+    }
+
+    private boolean isWithinRefreshThreshold(Long expirationTime) {
+        assert expirationTime != null;
+        return isWithinRefreshThreshold(Instant.now(), Instant.ofEpochSecond(expirationTime));
+    }
+
+    @VisibleForTesting
+    boolean isWithinRefreshThreshold(Instant currentInstant, Instant expiration) {
+        return currentInstant.plusSeconds(refreshThresholdInSeconds).getEpochSecond() >= expiration.getEpochSecond();
+    }
+
+    @VisibleForTesting
+    String refreshToken() {
+        long traceEnterId = LoggerHelpers.traceEnter(log, "refreshToken", this.scopeName, this.streamName);
+        CompletableFuture<Void> currentRefreshFuture = tokenRefreshFuture.get();
+        if (currentRefreshFuture == null) {
+            log.debug("Initiated token refresh for scope {} and stream {}", this.scopeName, this.streamName);
+            this.tokenRefreshFuture.set(this.recreateToken());
+        } else {
+            log.debug("Token is already under refresh for scope {} and stream {}", this.scopeName, this.streamName);
+        }
+        tokenRefreshFuture.get().join(); // Block until the token is refreshed
+        this.tokenRefreshFuture.compareAndSet(currentRefreshFuture, null); // Token is already refreshed, so resetting the future to null.
+
+        LoggerHelpers.traceLeave(log, "refreshToken", traceEnterId, this.scopeName, this.streamName);
+        return delegationToken.get().getValue();
```

Comment:
There is no point in a private method returning a member variable. The caller can access this themselves.

Reply:
Not returning the value makes the calling code `retrieveToken()` look overly complicated. I think we should leave it as is, as a result.